returns/interfaces/failable.py

```diff
@@ -0,0 +1,259 @@
+            container.from_value(raw_value),
+            container.from_value(raw_value).rescue(function),
+        )
+
+
+class FailableN(
+    ContainerN[_FirstType, _SecondType, _ThirdType],
+    RescuableN[_FirstType, _SecondType, _ThirdType],
+    Lawful['FailableN[_FirstType, _SecondType, _ThirdType]'],
+):
+    """
+    Base type for types that can fail.
+
+    It is a raw type and should not be used directly.
+    Use ``SingleFailableN`` and ``DiverseFailableN`` instead.
+    """
+
+    _laws: ClassVar[Sequence[Law]] = (
+        Law3(_FailableLawSpec.rescue_short_circuit_law),
+    )
+
+
+#: Type alias for kinds with two type arguments.
+Failable2 = FailableN[_FirstType, _SecondType, NoReturn]
+
+#: Type alias for kinds with three type arguments.
+Failable3 = FailableN[_FirstType, _SecondType, _ThirdType]
+
+
+@final
+class _SingleFailableLawSpec(LawSpecDef):
+    """
+    Single Failable laws.
+
+    We need to be sure that ``.map`` and ``.bind``
+    works correctly for ``empty`` property.
+    """
+
+    @law_definition
+    def map_short_circuit_law(
+        container: 'SingleFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot map from the `empty` property."""
+        assert_equal(
+            container.empty,
+            container.empty.map(function),
+        )
+
+    @law_definition
+    def bind_short_circuit_law(
+        container: 'SingleFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_FirstType],
+            KindN['SingleFailableN', _NewFirstType, _SecondType, _ThirdType],
+        ],
+    ) -> None:
+        """Ensures that you cannot bind from the `empty` property."""
+        assert_equal(
+            container.empty,
+            container.empty.bind(function),
+        )
+
+    @law_definition
+    def apply_short_circuit_law(
+        container: 'SingleFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot apply from the `empty` property."""
+        wrapped_function = container.from_value(function)
+        assert_equal(
+            container.empty,
+            container.empty.apply(wrapped_function),
+        )
+
+
+class SingleFailableN(
+    FailableN[_FirstType, _SecondType, _ThirdType],
+    Lawful['SingleFailableN[_FirstType, _SecondType, _ThirdType]'],
+):
+    """
+    Base type for types that have just only one failed value.
+
+    Like ``Maybe`` types where the only failed value is ``Nothing``.
+    """
+
+    _laws: ClassVar[Sequence[Law]] = (
+        Law2(_SingleFailableLawSpec.map_short_circuit_law),
+        Law2(_SingleFailableLawSpec.bind_short_circuit_law),
+        Law2(_SingleFailableLawSpec.apply_short_circuit_law),
+    )
+
+    @property
+    @abstractmethod
+    def empty(self) -> 'SingleFailableN[_FirstType, _SecondType, _ThirdType]':
```

Comment:
We need to annotate `self` here.

Reply:
https://github.com/dry-python/returns/blob/master/returns/interfaces/specific/reader.py#L97

Reply:
It didn't work 😢 😢

```python
class SingleFailableN(
    FailableN[_FirstType, _SecondType, _ThirdType],
    Lawful['SingleFailableN[_FirstType, _SecondType, _ThirdType]'],
):
    @property
    @abstractmethod
    def empty(
        self: _SingleFailableType,
    ) -> KindN[_SingleFailableType, _FirstType, _SecondType, _ThirdType]:
        """This property represents the failed value."""
```